### ISSUE TITLE
Issues in getting reasonable mAP while training

### ISSUE THREAD

**Issue by marcozov** [matterport/Mask_RCNN#1501]
Hi,

I am currently trying to train the model with the COCO dataset. In particular, I am using a modified version of the 2017 splits, which maintains the defined proportion (118k in the train split, 5k in the validation split), but it changes some of the elements inside each split (some of the images of the original COCO train split are moved in the validation and viceversa). Clearly, I have also defined new annotation/labels files, consistent with these new splits. I start the training by initializing the weights with imaginet: python coco.py train --dataset= MY_COCO_DIR --year=2017 --model=imagenet

I have this issue: the training does not seem to be very effective, since the mAP (computed through python coco.py evaluate --dataset=MY_COCO_DIR --year=2017 --model=MY_MASKRCNN_PATH/logs/coco20190502T2056/mask_rcnn_coco_0160.h5 ) is always quite low:

 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.050
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.127
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.030
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.022
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.058
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.084
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.073
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.098
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.099
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.032
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.092
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.151

This is obtained when the training is finished. I use a Titan X and the training takes around 4/5 days.

I am using a single GPU, therefore BATCH_SIZE=2 (since IMAGES_PER_GPU=2). I actually left all the configurations as the default ones, that’s why I having a hard time at figuring out why the model does not seem to properly fit the data. There are few lines that I took from the output, in case they can be useful:

> 

Epoch 1/40: 1000/1000 [==============================] - 1582s 2s/step - loss: 2.4615 - rpn_class_loss: 0.1257 - rpn_bbox_loss: 0.5467 - mrcnn_class_loss: 0.5672 - mrcnn_bbox_loss: 0.5331 - mrcnn_mask_loss: 0.6888 - val_loss: 3.8373 - val_rpn_class_loss: 0.2355 - val_rpn_bbox_loss: 1.1863 - val_mrcnn_class_loss: 0.7991 - val_mrcnn_bbox_loss: 0.9284 - val_mrcnn_mask_loss: 0.6879

Epoch 10/40: 1000/1000 [==============================] - 1480s 1s/step - loss: 1.5970 - rpn_class_loss: 0.0473 - rpn_bbox_loss: 0.3047 - mrcnn_class_loss: 0.3969 - mrcnn_bbox_loss: 0.3597 - mrcnn_mask_loss: 0.4885 - val_loss: 3.1093 - val_rpn_class_loss: 0.1345 - val_rpn_bbox_loss: 0.8850 - val_mrcnn_class_loss: 0.8571 - val_mrcnn_bbox_loss: 0.7166 - val_mrcnn_mask_loss: 0.5161

Epoch 61/120: 1000/1000 [==============================] - 1307s 1s/step - loss: 0.9952 - rpn_class_loss: 0.0181 - rpn_bbox_loss: 0.2322 - mrcnn_class_loss: 0.2095 - mrcnn_bbox_loss: 0.2122 - mrcnn_mask_loss: 0.3233 - val_loss: 2.5287 - val_rpn_class_loss: 0.0368 - val_rpn_bbox_loss: 1.2079 - val_mrcnn_class_loss: 0.4660 - val_mrcnn_bbox_loss: 0.4377 - val_mrcnn_mask_loss: 0.3803

Epoh 91/120: 1000/1000 [==============================] - 1285s 1s/step - loss: 0.9300 - rpn_class_loss: 0.0147 - rpn_bbox_loss: 0.2226 - mrcnn_class_loss: 0.2223 - mrcnn_bbox_loss: 0.1847 - mrcnn_mask_loss: 0.2857 - val_loss: 2.5603 - val_rpn_class_loss: 0.0723 - val_rpn_bbox_loss: 1.0890 - val_mrcnn_class_loss: 0.5659 - val_mrcnn_bbox_loss: 0.3655 - val_mrcnn_mask_loss: 0.4676

Epoch 121/160: 1000/1000 [==============================] - 1710s 2s/step - loss: 1.7273 - rpn_class_loss: 0.0469 - rpn_bbox_loss: 0.5513 - mrcnn_class_loss: 0.4245 - mrcnn_bbox_loss: 0.3350 - mrcnn_mask_loss: 0.3695 - val_loss: 2.2322 - val_rpn_class_loss: 0.1335 - val_rpn_bbox_loss: 0.8672 - val_mrcnn_class_loss: 0.3713 - val_mrcnn_bbox_loss: 0.4424 - val_mrcnn_mask_loss: 0.4179

Epoch 160/160: 1000/1000 [==============================] - 2213s 2s/step - loss: 1.3093 - rpn_class_loss: 0.0246 - rpn_bbox_loss: 0.3485 - mrcnn_class_loss: 0.3370 - mrcnn_bbox_loss: 0.2421 - mrcnn_mask_loss: 0.3571 - val_loss: 2.5502 - val_rpn_class_loss: 0.0882 - val_rpn_bbox_loss: 1.0248 - val_mrcnn_class_loss: 0.5397 - val_mrcnn_bbox_loss: 0.4790 - val_mrcnn_mask_loss: 0.4184

> 

Any idea about why is this happening and/or how to solve this problem? I have also tried to train the model with default 2014 split (without moving any image from one split to the other: python coco.py train --dataset=MY_COCO_PATH --year=2014 --model=imagenet --download=True) to find eventual bugs in my custom splits, but I have similar issues during the evaluation. I also tried to increase the learning rate but I start having a nan ( I guess that relates to the issue that you point out on the index page of the repository).

Thank you very much in advance, any help will be appreciated.

**Comment by koukouzasg** [comment]
@marcozov Hello, have you created the python notebooks? Inspect_model and inspect_weights? I found them really helpful and got a better view of my model after doing these. Also, try to do something similar with the demo notebook and visualize what your model produces. So in that way you can see if your model is performing bad or if you are actually using the mAP function wrong. :)

**Comment by spotiris** [comment]
It looks like your model is not converging to a low training or validation error.
As @koukouzasg mentioned, use the notebooks provided to inspect the data to ensure it is being fed to the network correctly.
If that fails, post your CONFIG and if possible a tensorboard plot of those losses, and someone may be able to help.
Are you training from scratch? I have had much more success training layers 4+ from ImageNet than training from scratch.

At an initial glance, your RPN is performing very poorly, causing the MRCNN heads to fail also.

**Comment by marcozov** [comment]
Thank you very much for your replies. By looking at the inspection_weights notebook, I compared the statistics of the model with my weights (the ones I got from the training of 160 epochs) to the model with the pre-trained weights (if things go well, theoretically, I think they should be pretty similar).

In the "Review Weight Stats" tab, "bn_conv1/moving_variance:0" gets a very high value for both of the setups (i.e., pre-trained weights and my weights after 160 epochs): in both cases a warning about overflow is shown.

Then, when I examine the entry "bn4g_branch2a/moving_mean:0" and "bn4g_branch2a/moving_variance:0" I see that they have value 0 and 1, respectively, and warning denoted as "dead?". I guess it means that the related weight is not getting updated through the iterations. This happens in many other cases: ("bn4g_branch2b/moving_mean:0", bn4g_branch2b/moving_variance:0), ("bn4g_branch2c/moving_mean:0", "bn4g_branch2c/moving_variance:0") etc. .

Also, there is another (I guess related? ) issue when I look at the weights' histograms: for some weights, their distribution seems to be uniform inside a certain interval. For instance: res4g_branch2a/kernel:0, res4g_branch2b/kernel:0 among others. Of course, I am talking about the model initialized with my weights.

I am attaching the statistics that I have been discussing as an html file (containing the table with the warnings, but for attaching it I needed to rename it as .txt so you need to remove this extension in order to have an html file that can be opened with the browser) and an image (for the histograms).

I am a bit unexperienced in debugging deep models, so I don't really know how what these statistics express and how to interpret them. Thanks again.
[weight_stats.html.txt](https://github.com/matterport/Mask_RCNN/files/3192215/weight_stats.html.txt)

![histogramOfWeights](https://user-images.githubusercontent.com/17951561/57940814-4d150d00-789b-11e9-8e6e-bcbc1555535b.png)
